Interpreter/ProgramTree/Nodes/BodyNode.java

```diff
@@ -16,9 +19,19 @@ public BodyNode() {
     public static BodyNode parseNode(TokenStack tokens) {
         tokens.pushStack();
 
-        
+        BodyNode body = new BodyNode();
+        BodyStatementNodeBase statement;
+
+        while (tokens.peekToken().getTokenType() != TokenType.R_BRACE) {
+            statement = BodyStatementNodeBase.parseNode(tokens);
+            if (statement == null) {
+                tokens.popStack(true);
+                return null;
+            }
+            body.statements.add(statement);
+        }
 
```

Comment:
Not quite. This will never pop the closing brace, and the next node's parsing will fail.

Also, this is just a preference, but I'd pass the `ArrayList `into the constructor and not create our `BodyNode `instance until we return it. I prefer not to access private members unless you're doing it from the instance or its unavoidable like a comparison method. As a minor additional benefit, that's one less object for the garbage for collector (but that's doesn't really matter here).